bluetooth: fast_pair: use_case: locator_tag: introduce fp_adv module

kapi-no opened the pull request:
Added the new Fast Pair advertising module to the locator tag use case layer of the Google Fast Pair subsystem. The new module is used by the Fast Pair Locator Tag sample.

Ref: NCSDK-30487

NordicBuilder commented:
## CI Information

<sup>_To view the history of this post, clich the 'edited' button above_</sup>
Build number: 1
### Inputs:
#### Sources:
sdk-nrf: PR head: [a22014655569779a775e73d406e5a7c8dc1ccf32](https://github.com/nrfconnect/sdk-nrf/pull/19571/commits/a22014655569779a775e73d406e5a7c8dc1ccf32)
<details><summary> more details</summary>

sdk-nrf:
> PR head: [a22014655569779a775e73d406e5a7c8dc1ccf32](https://github.com/nrfconnect/sdk-nrf/pull/19571/commits/a22014655569779a775e73d406e5a7c8dc1ccf32)
> merge base: [1536c9082ff9e6a2105ddc2b85a8c0928c8e458a](https://github.com/nrfconnect/sdk-nrf/commit/1536c9082ff9e6a2105ddc2b85a8c0928c8e458a)
> target head (main): [b0461ac723c3e66f35c829a62223f9cbc2ea34d0](https://github.com/nrfconnect/sdk-nrf/commit/b0461ac723c3e66f35c829a62223f9cbc2ea34d0)
> [Diff](https://github.com/nrfconnect/sdk-nrf/compare/1536c9082ff9e6a2105ddc2b85a8c0928c8e458a..a22014655569779a775e73d406e5a7c8dc1ccf32)</details>

#### Github labels
| Enabled | Name | Description |
| -- | -- | -- |
|| ci-disabled | _Disable the ci execution_ | 
|| ci-all-test | _Run all of ci, no test spec filtering will be done_ | 
|| ci-force-downstream | _Force execution of downstream even if twister fails_ | 
|| ci-run-twister | _Force run twister_ | 
|| ci-run-zephyr-twister | _Force run zephyr twister_ | 

<details><summary> List of changed files detected by CI (16)</summary>

```
include
│  ├── bluetooth
│  │  ├── services
│  │  │  ├── fast_pair
│  │  │  │  ├── use_case
│  │  │  │  │  ├── locator_tag
│  │  │  │  │  │  │ fp_adv.h
samples
│  ├── bluetooth
│  │  ├── fast_pair
│  │  │  ├── locator_tag
│  │  │  │  ├── CMakeLists.txt
│  │  │  │  ├── configuration
│  │  │  │  │  ├── prj.conf
│  │  │  │  │  │ prj_release.conf
│  │  │  │  ├── include
│  │  │  │  │  │ app_fp_adv.h
│  │  │  │  ├── src
│  │  │  │  │  ├── dfu.c
│  │  │  │  │  ├── factory_reset.c
│  │  │  │  │  ├── fp_adv_trigger.ld
│  │  │  │  │  │ main.c
subsys
│  ├── bluetooth
│  │  ├── services
│  │  │  ├── fast_pair
│  │  │  │  ├── CMakeLists.txt
│  │  │  │  ├── use_case
│  │  │  │  │  ├── CMakeLists.txt
│  │  │  │  │  ├── Kconfig
│  │  │  │  │  ├── locator_tag
│  │  │  │  │  │  ├── CMakeLists.txt
│  │  │  │  │  │  ├── Kconfig
│  │  │  │  │  │  ├── fp_adv.c
│  │  │  │  │  │  │ fp_adv_trigger.ld
```
</details>

### Outputs:

#### Toolchain
Version: 
Build docker image: 

#### Test Spec & Results: :white_check_mark: Success; :x: Failure; :orange_circle: Queued; :yellow_circle: Progress; :white_medium_square: Skipped; :warning: Quarantine
- :orange_circle: Toolchain
- :orange_circle: Build twister
- :orange_circle: Integration tests
   - :orange_circle: desktop52_verification
   - :orange_circle: test_ble_nrf_config
   - :orange_circle: test-fw-nrfconnect-ble_samples
   - :orange_circle: test-sdk-find-my

<details><summary>Disabled integration tests</summary>

- 
   - doc-internal
   - test-fw-nrfconnect-apps
   - test-fw-nrfconnect-ble_mesh
   - test-fw-nrfconnect-boot
   - test-fw-nrfconnect-chip
   - test-fw-nrfconnect-fem
   - test-fw-nrfconnect-nfc
   - test-fw-nrfconnect-nrf-iot_libmodem-nrf
   - test-fw-nrfconnect-nrf-iot_lwm2m
   - test-fw-nrfconnect-nrf-iot_mosh
   - test-fw-nrfconnect-nrf-iot_nrf_provisioning
   - test-fw-nrfconnect-nrf-iot_positioning
   - test-fw-nrfconnect-nrf-iot_samples
   - test-fw-nrfconnect-nrf-iot_serial_lte_modem
   - test-fw-nrfconnect-nrf-iot_thingy91
   - test-fw-nrfconnect-nrf-iot_zephyr_lwm2m
   - test-fw-nrfconnect-nrf_crypto
   - test-fw-nrfconnect-ps
   - test-fw-nrfconnect-rpc
   - test-fw-nrfconnect-rs
   - test-fw-nrfconnect-tfm
   - test-fw-nrfconnect-thread
   - test-fw-nrfconnect-zigbee
   - test-low-level
   - test-sdk-audio
   - test-sdk-dfu
   - test-sdk-mcuboot
   - test-sdk-pmic-samples
   - test-sdk-sidewalk
   - test-sdk-wifi
   - test-secdom-samples-public
</details>



*Note: This message is automatically posted and updated by the CI*

kapi-no commented on include/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.h:
```diff
@@ -0,0 +1,183 @@
+
+/**
+ * @defgroup bt_fast_pair_locator_tag_fp_adv Fast Pair Locator Tag advertising API
+ * @brief Fast Pair Locator Tag API for managaing the Fast Pair advertising set
```
thanks, fixed

MarekPieta commented on subsys/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.c:
```diff
@@ -722,22 +720,49 @@ int app_fp_adv_disable(void)
 	err = fp_adv_set_teardown();
 	if (err) {
-		LOG_ERR("Fast Pair: fp_adv_set_teardown failed (err %d)", err);
+		LOG_ERR("Fast Pair Locator Tag FP Adv: fp_adv_set_teardown failed (err %d)", err);
```
Should we keep `is_enabled=false` even if the advertising set teardown fails? It could lead to problems

kapi-no replied:
I will improve to make it work similar to the FMDN state module.

MarekPieta commented on include/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.h:
```diff
@@ -0,0 +1,183 @@
+ *
+ *  @return true when the module is ready, false otherwise.
+ */
+bool bt_fast_pair_locator_tag_fp_adv_is_ready(void);
```
I also wonder if we could make the module automatically react on the factory reset (it could even clear NVM data related to the used Bluetooth identity)

MarekPieta replied:
Discussed. We will keep it as convenience API.

MarekPieta commented on subsys/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.c:
```diff
@@ -161,20 +163,20 @@ static bool can_pair(void)
+static bool fp_adv_is_pairing_mode(enum bt_fast_pair_locator_tag_fp_adv_mode fp_adv_mode)
 {
-	return ((fp_adv_mode == APP_FP_ADV_MODE_DISCOVERABLE) && can_pair());
+	return ((fp_adv_mode == BT_FAST_PAIR_LOCATOR_TAG_FP_ADV_MODE_DISCOVERABLE) && can_pair());
```
Discussed offline. We will use `_adv_mgmt_` infix for the new functionality

kapi-no replied:
Changed to use `_adv_manager_` to be consistent with the current naming convention (BT_FAST_PAIR_BOND_MANAGER / BT_FAST_PAIR_STORAGE_MANAGER).

MarekPieta commented on include/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.h:
```diff
@@ -0,0 +1,183 @@
+ *  Use with caution, as it updates the Fast Pair advertising payload asynchronously
+ *  to the RPA address rotation and to the update of the FMDN advertising payload.
+ */
+void bt_fast_pair_locator_tag_fp_adv_payload_refresh(void);
```
Discussed offline. We could use a `structure` instead of `boolean` for `force_payload_refresh` to avoid breaking API in future.

kapi-no replied:
Decided to leave it as a standalone function as it may also be used in combination with the `bt_fast_pair_adv_manager_mode_set` API.

MarekPieta commented on include/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.h:
```diff
@@ -0,0 +1,183 @@
+ *
+ *  @return 0 if the operation was successful. Otherwise, a (negative) error code is returned.
+ */
+int bt_fast_pair_locator_tag_fp_adv_enable(void);
```
Discussed offline. We do not want to do it to maintain control over advertising set allocation time

MarekPieta commented on include/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.h:
```diff
@@ -0,0 +1,183 @@
+ *
+ *  @return 0 if the operation was successful. Otherwise, a (negative) error code is returned.
+ */
+int bt_fast_pair_locator_tag_fp_adv_id_set(uint8_t id);
```
Discussed offline. We will keep setter to maintain the same approach as for other existing applications.

MarekPieta commented on include/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.h:
```diff
@@ -0,0 +1,183 @@
+#endif
+
+/** Fast Pair Locator Tag advertising mode. */
+enum bt_fast_pair_locator_tag_fp_adv_mode {
```
Discussed offline. We will rely on additional boolean

kapi-no commented on include/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.h:
```diff
@@ -0,0 +1,183 @@
+ */
+bool bt_fast_pair_locator_tag_fp_adv_is_ready(void);
+
+/** @brief Initialize the Fast Pair advertising module.
```
removed to simplify API

kapi-no commented on subsys/bluetooth/services/fast_pair/use_case/locator_tag/fp_adv.c:
```diff
@@ -196,14 +198,15 @@ static int fp_adv_payload_set(bool rpa_rotated, bool new_session)
 	err = bt_hci_get_adv_handle(fp_adv_set, &adv_handle);
 	if (err) {
-		LOG_ERR("Fast Pair: cannot get advertising handle (err: %d)", err);
+		LOG_ERR("Fast Pair Locator Tag FP Adv: cannot get advertising handle (err: %d)",
```
Kept it to keep convention with the remaining Fast Pair modules. We have an old task to migrate to the logging prefixes in the backlog

kapi-no commented:
Replaced by the following PR after I changed the approach (I plan to make this module more general):

https://github.com/nrfconnect/sdk-nrf/pull/19800
